Exercici 1

In [1]:
# Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe.
# Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris
# rellevants. Justifica la teva elecció.

import pandas as pd

pd.read_csv('DelayedFlights.csv')

Airlines_Delay = pd.read_csv('DelayedFlights.csv')

pd.options.display.max_columns = None

Airlines_Delay


Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0         2211.0        2225            WN        335  N712SW   
1         1002.0        1000            WN       3231  N772SW   
2          804.0         750            WN        448  N428WN   
3         1959.0        1925            WN       3920  N464WN   
4         2121.0        2110            WN        378  N726SW   
...          ...         ...           ...        ...     ...   
1936753   1617.0        1552            DL       1621  N938DL   
1936754    904.0         749            DL       1631  N3743H   
1936755   1149.0        1010            DL       1631  N909DA   
1936756   1446.0        1437            DL       1639  N646DL   
1936757   1413.0        1418            DL       1641  N908DL   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
0                    128.0           150.0    116.0     -14.0       8.0   
1                    128.0           145.0    113.0       2.0      19.0   
2                     96.0            90.0     76.0      14.0       8.0   
3                     90.0            90.0     77.0      34.0      34.0   
4                    101.0           115.0     87.0      11.0      25.0   
...                    ...             ...      ...       ...       ...   
1936753              147.0           152.0    120.0      25.0      30.0   
1936754              127.0           109.0     78.0      75.0      57.0   
1936755              162.0           143.0    122.0      99.0      80.0   
1936756              115.0           117.0     89.0       9.0      11.0   
1936757              123.0           135.0    104.0      -5.0       7.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
0          IAD  TPA       810     4.0      8.0          0                N   
1          IAD  TPA       810     5.0     10.0          0                N   
2          IND  BWI       515     3.0     17.0          0                N   
3          IND  BWI       515     3.0     10.0          0                N   
4          IND  JAX       688     4.0     10.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
1936753    MSP  ATL       906     9.0     18.0          0                N   
1936754    RIC  ATL       481    15.0     34.0          0                N   
1936755    ATL  IAH       689     8.0     32.0          0                N   
1936756    IAD  ATL       533    13.0     13.0          0                N   
1936757    SAT  ATL       874     8.0     11.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0               0           NaN           NaN       NaN            NaN   
1               0           NaN           NaN       NaN            NaN   
2               0           NaN           NaN       NaN            NaN   
3               0           2.0           0.0       0.0            0.0   
4               0           NaN           NaN       NaN    

# Variables

## Identificaió del registre
- Primera columna: posició del registre

## Data i horari del vol
- Year: Any en que s'ha realitzat el vol
- Month: Mes en que s'ha realitzat el vol
- DayofMonth: Día del mes en que s'ha realitzat el vol
- DayOfWeeK: Día de la setmana en que s'ha realitzat el vol (nº)
- DepTime: Hora de sortida actual
- CRSDepTime: Hora de sortida planificada
- ArrTime: Hora d'arribada actual
- CRSArrTime: Hora d'arribada planificada

## Identificació de l'aerolinia, vol i avió
- UniqueCarrier: codi únic de l'aerolínia
- FlightNum: número de vol
- TailNum: identificació de l'avió (matrícula)

## Temps de vol
- ActualElapsedTime: temps actual del vol entre porta de sortida i porta d'arribada (minuts)
- CRSElapsedTime: temps estimat del vol entre porta de sortida i porta d'arribada (minuts)
- AirTime: temps de vol (minuts)
- ArrDelay: Diferencia en minuts entre l'hora estimada d'arribada i la real (si el vol arriba abans de l'hora estimada, número negatiu en minuts)
- DepDelay: Diferencia en minuts entre l'hora estimada de sortida i la real (si el vol surt abans de l'hora estimada, número negatiu en minuts)

## Ruta del vol
- Origin: Codi IATA de l'aeroport d'origen
- Dest: Codi IATA de l'aeroport de destí
- Distance: Distancia entre aeroports (milles)

## Temps de les operacions d'enlairament i d'aterratge
- TaxiIn: temps desde aterratge fins arribada a la porta d'embarcament (minuts)
- TaxiOut: temps desde porta d'embarcament fins l'enlairament (minuts)

## Cancel·lació, codi de cancel·lació i desviació
- Cancelled: Si el vol s'ha cancel·lat o no. 1 = Sí, 0 = No
- CancellationCode: codi segons causa de cancel·lació. (A = Aerolinia, B = Metereologia extrema, C = Sistema Nacional d'Aviació (tràfic aeri, meteorologia no extrema, operacions aeroportuaries, etc.), D = Seguretat)
- Diverted: Si el vol s'ha desviat o no. 1 = Sí, 0 = No

## Retards
- CarrierDelay:Retard aerolinia (minuts) (NaN: Not a Number)
- WeatherDelay: Retard meteorologia (minuts)
- NASDelay: Retard Sistema Nacional d'Aviació (minuts)
- SecurityDelay: Retard Seguretat (minuts)
- LateAircraftDelay: Retard per el retard del vol anterior (minuts)

# Elecció variables

## Per al anàlisis dels retards i les seves causes, de cada una de les aerolinies, es consideren necessaries les variables següents:

### Identificaió del registre 
### Data i horari del vol: totes les variables.
### Identificació de l'aerolinia, vol i avió: la variable ***UniqueCarrier***
### Temps de vol: les variables ***ActualElapsedTime*** , ***CRSElapsedTime*** , ***AirTime***, ***ArrDelay***, ***DepDelay***
### Ruta del vol: les variables ***Origin*** , ***Dest*** i ***Distance*** (amb la distancia i el temps de vol (aire) podria calcular la velocitat del vol)
### Temps de les operacions d'enlairament i d'aterratge: cap variable
### Cancel·lació, codi de cancel·lació i desviació: la variable ***CancellationCode*** 
### Retards: totes les variables


<font size = 4.5> Amb aquestes variables podem determinar si una vol a sofert retard i les seves causes, identificant vol per data, horari, ruta i aerolinia.

## Les següents variables no serán considerades:

### ***TailNum***. Crec que identificant el número de vol i la aerolinia, ja aporta suficient informació
### Respecte al temp de vol, la variable ***ArrDelay***. Disposant del temps total de vol, actual i estimat, entre porta de sortida i porta d'arribada, podem veure l'afectació del retard.
### Respecte a les operacions  d'enlairament i d'aterratge, les variables ***TaxiIn*** i ***TaxiOut***. Crec que aquestes dues variables ens servirien per analitzar les operacions propies del aeroport.
### Respecte a les variables de cancel·lació, la varible ***Cancelled***. Disposant del codi del ***CancellationCode***, ja determinem si el vol s'ha retrassat o no.



In [2]:
# Dataframe amb variables considerades rellevants

import pandas as pd

Delayed_Flights_Variables = pd.read_csv('DelayedFlights.csv')

Delayed_Flights_Variables.drop(['Unnamed: 0', 'TailNum', 'TaxiIn', 'TaxiOut', 'Cancelled', 'Diverted'], axis = 1)


Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
3        2008      1           3          4   1829.0        1755   1959.0   
4        2008      1           3          4   1940.0        1915   2121.0   
...       ...    ...         ...        ...      ...         ...      ...   
1936753  2008     12          13          6   1250.0        1220   1617.0   
1936754  2008     12          13          6    657.0         600    904.0   
1936755  2008     12          13          6   1007.0         847   1149.0   
1936756  2008     12          13          6   1251.0        1240   1446.0   
1936757  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ActualElapsedTime  \
0              2225            WN        335              128.0   
1              1000            WN       3231              128.0   
2               750            WN        448               96.0   
3              1925            WN       3920               90.0   
4              2110            WN        378              101.0   
...             ...           ...        ...                ...   
1936753        1552            DL       1621              147.0   
1936754         749            DL       1631              127.0   
1936755        1010            DL       1631              162.0   
1936756        1437            DL       1639              115.0   
1936757        1418            DL       1641              123.0   

         CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  \
0                 150.0    116.0     -14.0       8.0    IAD  TPA       810   
1                 145.0    113.0       2.0      19.0    IAD  TPA       810   
2                  90.0     76.0      14.0       8.0    IND  BWI       515   
3                  90.0     77.0      34.0      34.0    IND  BWI       515   
4                 115.0     87.0      11.0      25.0    IND  JAX       688   
...                 ...      ...       ...       ...    ...  ...       ...   
1936753           152.0    120.0      25.0      30.0    MSP  ATL       906   
1936754           109.0     78.0      75.0      57.0    RIC  ATL       481   
1936755           143.0    122.0      99.0      80.0    ATL  IAH       689   
1936756           117.0     89.0       9.0      11.0    IAD  ATL       533   
1936757           135.0    104.0      -5.0       7.0    SAT  ATL       874   

        CancellationCode  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                      N           NaN           NaN       NaN            NaN   
1                      N           NaN           NaN       NaN            NaN   
2                      N           NaN           NaN       NaN            NaN   
3                      N           2.0           0.0       0.0            0.0   
4                      N           NaN           NaN       NaN            NaN   
...                  ...           ...           ...       ...            ...   
1936753                N           3.0           0.0       0.0            0.0   
1936754                N           0.0          57.0      18.0            0.0   
1936755                N           1.0           0.0      19.0            0.0   
1936756                N           NaN           NaN       NaN            NaN   
1936757                N           NaN           NaN       NaN            NaN   

         LateAircraftDelay  
0                      NaN  
1                      NaN  
2                      NaN  
3                     32.0  
4                      NaN  
...                    ...  
1936753               22.0  
1936754                0.0  
1936755               79.0  
1936756                NaN  
1936757                NaN  

[1936758 rows x 24 c

In [4]:
# Del Dataframe amb les variables seleccionades "Delayed_Flights_Variables", el reduïm aleatoriament a 200.000 registres

import pandas as pd

Delayed_Flights_Variables = pd.read_csv('DelayedFlights.csv')

Delayed_Flights_Variables.drop(['Unnamed: 0', 'TailNum', 'TaxiIn', 'TaxiOut', 'Cancelled', 'Diverted'], axis = 1)

# reduïm aleatoriament a 200.000 registres

Delayed_Flights_Reduit = Delayed_Flights_Variables.sample(n=200000)

pd.options.display.max_columns = None

Delayed_Flights_Reduit



Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
224054       711973  2008      2          18          1    734.0         720   
1775762     6562647  2008     12          23          2    901.0         835   
1095290     3643711  2008      7          14          1   1202.0        1120   
185663       610669  2008      2           4          1   1620.0        1525   
173677       566387  2008      1           3          4     40.0        2215   
...             ...   ...    ...         ...        ...      ...         ...   
160883       521433  2008      1          31          4   1257.0        1250   
559233      1749227  2008      3          17          1   1121.0        1115   
1087354     3620401  2008      7           7          1   2112.0        2100   
977672      3274068  2008      6          16          1   2235.0        2220   
359589      1130799  2008      2          24          7   1024.0        1015   

         ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
224054    1040.0        1047            XE       2894  N14920   
1775762   1014.0         955            XE       2941  N12934   
1095290   1212.0        1140            WN       1912  N349SW   
185663    2337.0        2255            WN        446  N237WN   
173677     857.0         629            CO        703  N73299   
...          ...         ...           ...        ...     ...   
160883    1542.0        1515            AA       1971  N459AA   
559233    1414.0        1420            CO        438  N14653   
1087354   2216.0        2210            WN       3207  N613SW   
977672    2336.0        2322            US       1788  N454UW   
359589    1316.0        1303            B6        540  N562JB   

         ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay  \
224054               126.0           147.0    113.0      -7.0      14.0   
1775762               73.0            80.0     55.0      19.0      26.0   
1095290               70.0            80.0     57.0      32.0      42.0   
185663               257.0           270.0    236.0      42.0      55.0   
173677               317.0           314.0    290.0     148.0     145.0   
...                    ...             ...      ...       ...       ...   
160883               225.0           205.0    198.0      27.0       7.0   
559233               173.0           185.0    144.0      -6.0       6.0   
1087354               64.0            70.0     53.0       6.0      12.0   
977672               121.0           122.0     99.0      14.0      15.0   
359589               172.0           168.0    141.0      13.0       9.0   

        Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  \
224054     MSY  CLE       917     6.0      7.0          0                N   
1775762    IAH  BRO       308     4.0     14.0          0                N   
1095290    SLC  LAS       368     4.0      9.0          0                N   
185663     LAS  ORF      2155     5.0     16.0          0                N   
173677     SJC  EWR      2548    12.0     15.0          0                N   
...        ...  ...       ...     ...      ...        ...              ...   
160883     DCA  DFW      1192     8.0     19.0          0                N   
559233     EWR  MIA      1086     7.0     22.0          0                N   
1087354    STL  TUL       351     3.0      8.0          0                N   
977672     CLT  ORD       599     5.0     17.0          0                N   
359589     PBI  EWR      1024     8.0     23.0          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
224054          0           NaN           NaN       NaN            NaN   
1775762         0          19.0           0.0       0.0            0.0   
1095290         0           0.0           0.0       0.0            0.0   
185663          0          40.0           0.0       0.0            0.0   
173677          0           0.0           0.0       3.0    

In [7]:
# Salvem el nou Dataframe com a dataset "DelayedFlights_Reduit.csv"

Delayed_Flights_Reduit.to_csv('DelayedFlights_Reduit.csv')

----------------------------------------------------------------------------------------------

Exercici 2

In [8]:
# Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit.csv')

DelayedFlights


Unnamed: 0.1  Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  \
0             224054      711973  2008      2          18          1    734.0   
1            1775762     6562647  2008     12          23          2    901.0   
2            1095290     3643711  2008      7          14          1   1202.0   
3             185663      610669  2008      2           4          1   1620.0   
4             173677      566387  2008      1           3          4     40.0   
...              ...         ...   ...    ...         ...        ...      ...   
199995        160883      521433  2008      1          31          4   1257.0   
199996        559233     1749227  2008      3          17          1   1121.0   
199997       1087354     3620401  2008      7           7          1   2112.0   
199998        977672     3274068  2008      6          16          1   2235.0   
199999        359589     1130799  2008      2          24          7   1024.0   

        CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0              720   1040.0        1047            XE       2894  N14920   
1              835   1014.0         955            XE       2941  N12934   
2             1120   1212.0        1140            WN       1912  N349SW   
3             1525   2337.0        2255            WN        446  N237WN   
4             2215    857.0         629            CO        703  N73299   
...            ...      ...         ...           ...        ...     ...   
199995        1250   1542.0        1515            AA       1971  N459AA   
199996        1115   1414.0        1420            CO        438  N14653   
199997        2100   2216.0        2210            WN       3207  N613SW   
199998        2220   2336.0        2322            US       1788  N454UW   
199999        1015   1316.0        1303            B6        540  N562JB   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
0                   126.0           147.0    113.0      -7.0      14.0    MSY   
1                    73.0            80.0     55.0      19.0      26.0    IAH   
2                    70.0            80.0     57.0      32.0      42.0    SLC   
3                   257.0           270.0    236.0      42.0      55.0    LAS   
4                   317.0           314.0    290.0     148.0     145.0    SJC   
...                   ...             ...      ...       ...       ...    ...   
199995              225.0           205.0    198.0      27.0       7.0    DCA   
199996              173.0           185.0    144.0      -6.0       6.0    EWR   
199997               64.0            70.0     53.0       6.0      12.0    STL   
199998              121.0           122.0     99.0      14.0      15.0    CLT   
199999              172.0           168.0    141.0      13.0       9.0    PBI   

       Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  \
0       CLE       917     6.0      7.0          0                N         0   
1       BRO       308     4.0     14.0          0                N         0   
2       LAS       368     4.0      9.0          0                N         0   
3       ORF      2155     5.0     16.0          0                N         0   
4       EWR      2548    12.0     15.0          0                N         0   
...     ...       ...     ...      ...        ...              ...       ...   
199995  DFW      1192     8.0     19.0          0                N         0   
199996  MIA      1086     7.0     22.0          0                N         0   
199997  TUL       351     3.0      8.0          0                N         0   
199998  ORD       599     5.0     17.0          0                N         0   
199999  EWR      1024     8.0     23.0          0                N         0   

        CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
0                NaN           NaN       NaN            NaN                NaN  
1               19.0      

In [9]:
# Informació bàscia del dataframe

DelayedFlights.info()

DelayedFlights.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 31 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0.1       200000 non-null  int64  
 1   Unnamed: 0         200000 non-null  int64  
 2   Year               200000 non-null  int64  
 3   Month              200000 non-null  int64  
 4   DayofMonth         200000 non-null  int64  
 5   DayOfWeek          200000 non-null  int64  
 6   DepTime            200000 non-null  float64
 7   CRSDepTime         200000 non-null  int64  
 8   ArrTime            199229 non-null  float64
 9   CRSArrTime         200000 non-null  int64  
 10  UniqueCarrier      200000 non-null  object 
 11  FlightNum          200000 non-null  int64  
 12  TailNum            200000 non-null  object 
 13  ActualElapsedTime  199099 non-null  float64
 14  CRSElapsedTime     199981 non-null  float64
 15  AirTime            199099 non-null  float64
 16  Ar

Unnamed: 0.1    Unnamed: 0      Year          Month     DayofMonth  \
count  2.000000e+05  2.000000e+05  200000.0  200000.000000  200000.000000   
mean   9.666874e+05  3.335452e+06    2008.0       6.099715      15.750725   
std    5.592476e+05  2.066021e+06       0.0       3.482171       8.787656   
min    1.100000e+01  1.800000e+01    2008.0       1.000000       1.000000   
25%    4.812825e+05  1.509636e+06    2008.0       3.000000       8.000000   
50%    9.668840e+05  3.239060e+06    2008.0       6.000000      16.000000   
75%    1.450894e+06  4.965649e+06    2008.0       9.000000      23.000000   
max    1.936756e+06  7.009726e+06    2008.0      12.000000      31.000000   

           DayOfWeek        DepTime     CRSDepTime        ArrTime  \
count  200000.000000  200000.000000  200000.000000  199229.000000   
mean        3.984600    1517.609950    1467.323380    1610.565515   
std         1.997276     450.420896     424.155354     546.628551   
min         1.000000       1.000000       0.000000       1.000000   
25%         2.000000    1203.000000    1135.000000    1316.000000   
50%         4.000000    1545.000000    1510.000000    1715.000000   
75%         6.000000    1900.000000    1815.000000    2030.000000   
max         7.000000    2400.000000    2359.000000    2400.000000   

          CRSArrTime      FlightNum  ActualElapsedTime  CRSElapsedTime  \
count  200000.000000  200000.000000      199099.000000   199981.000000   
mean     1633.581325    2186.217265         133.277214      134.241338   
std       463.860327    1945.206868          72.162271       71.371263   
min         1.000000       1.000000          14.000000        1.000000   
25%      1325.000000     612.000000          80.000000       81.000000   
50%      1705.000000    1543.000000         116.000000      116.000000   
75%      2013.000000    3431.000000         165.000000      165.000000   
max      2400.000000    9740.000000        1114.000000      660.000000   

             AirTime       ArrDelay       DepDelay       Distance  \
count  199099.000000  199099.000000  200000.000000  200000.000000   
mean      108.228002      42.083777      43.009170     764.962195   
std        68.714151      56.266634      52.786571     574.577518   
min         0.000000     -59.000000       6.000000      11.000000   
25%        58.000000       9.000000      12.000000     338.000000   
50%        90.000000      24.000000      24.000000     606.000000   
75%       137.000000      56.000000      53.000000     998.000000   
max      1091.000000    1279.000000    1280.000000    4962.000000   

              TaxiIn        TaxiOut      Cancelled       Diverted  \
count  199229.000000  199967.000000  200000.000000  200000.000000   
mean        6.813923      18.254117       0.000270       0.004235   
std         5.263404      14.506144       0.016429       0.064939   
min         0.000000       0.000000       0.000000       0.000000   
25%         4.000000      10.000000       0.000000       0.000000   
50%         6.000000      14.000000       0.000000       0.000000   
75%         8.000000      21.000000       0.000000       0.000000   
max       197.000000     342.000000       1.000000       1.000000   

        CarrierDelay   WeatherDelay       NASDelay  SecurityDelay  \
count  128700.000000  128700.000000  128700.000000  128700.000000   
mean       19.135789       3.651368      14.972821       0.088632   
std        43.063082      20.769977      33.759061       1.932348   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         2.000000       0.000000       2.000000       0.000000   
75%        21.000000       0.000000      15.000000       0.000000   
max      1198.000000     907.000000     914.000000     211.000000   

       LateAircraftDelay  
count      128700.000000  
mean           25.311943  
std            41.735405  
min             0.000000  
25%             0.00000

*----------------------------------------------------------------------------------------------------------------------------*

In [50]:
# Fes una anàlisi estadístic del que consideris rellevant.

 1. Anàlisi de la diferencia (minuts) entre l'hora de sortida, estimada i real, per aerolínies, per determinar quina aerolínia acumula més retrasos en la sortida

In [10]:
# Data frame amb columnes "UniqueCarrier", "DepDelay"


import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit.csv')

Diferencia_hora_sortida = DelayedFlights[['UniqueCarrier','DepDelay']]

Diferencia_hora_sortida.sort_values(['UniqueCarrier', 'DepDelay'])



UniqueCarrier  DepDelay
1765              9E       6.0
2081              9E       6.0
2284              9E       6.0
2372              9E       6.0
4173              9E       6.0
...              ...       ...
77647             YV     405.0
189903            YV     408.0
39354             YV     418.0
117366            YV     509.0
188606            YV     526.0

[200000 rows x 2 columns]

In [11]:
# Sumem las diferencies en minuts per companyia aerolinia

Diferencia_hora_sortida.groupby('UniqueCarrier').sum()

DepDelay
UniqueCarrier           
9E              262676.0
AA              923477.0
AQ                2617.0
AS              150834.0
B6              312925.0
CO              451537.0
DL              458161.0
EV              395930.0
F9               80607.0
FL              318699.0
HA               21311.0
MQ              643130.0
NW              324274.0
OH              254227.0
OO              610761.0
UA              728044.0
US              405043.0
WN             1343663.0
XE              529284.0
YV              384634.0

In [12]:
# Ordenem els retrasos acumalats per aerolinea de major a menor.

Diferencia_hora_sortida.groupby('UniqueCarrier').sum().sort_values('DepDelay', ascending = False)

DepDelay
UniqueCarrier           
WN             1343663.0
AA              923477.0
UA              728044.0
MQ              643130.0
OO              610761.0
XE              529284.0
DL              458161.0
CO              451537.0
US              405043.0
EV              395930.0
YV              384634.0
NW              324274.0
FL              318699.0
B6              312925.0
9E              262676.0
OH              254227.0
AS              150834.0
F9               80607.0
HA               21311.0
AQ                2617.0

A priori, havent analitzat la variable ***DepDelay*** del dataframe de 200.000 resgitres, s'observa que la aerolínea que acumula més retrasos en la sortida del vols, és la aerolínea WN i la que menys retrasos acomula és la aerolínea AQ. No obstannt caldrà anlitzar i contrastar-ho amb el número de vols realitzat per cada aerolínia.

In [13]:
# Calculem el número de vols per aerolínea

Diferencia_hora_sortida[('UniqueCarrier')].value_counts()

WN    38803
AA    19835
MQ    14812
UA    14583
OO    13602
DL    11850
XE    10604
CO    10509
US    10345
EV     8414
NW     8120
FL     7307
YV     7017
B6     5623
9E     5515
OH     5351
AS     3989
F9     2921
HA      715
AQ       85
Name: UniqueCarrier, dtype: int64

Havent calculat el número de vols per aerolínea, dels registres que tenim al dataframe, observem que efectivament la aerolínia que acumula menys retrasos és la que menys vols te registrats i la que mes retrasos acumula, és la que te més vols regitrats. Per tant l'anàlisis inicial no sería del tot correcte. Mirem de calcular la mitjana dels minuts acumulats entre les hores de sortida estimades i reals per aerolínia.

In [14]:
Diferencia_hora_sortida.groupby('UniqueCarrier').median().sort_values('DepDelay')

DepDelay
UniqueCarrier          
AQ                 16.0
HA                 16.0
F9                 17.0
WN                 20.0
AS                 20.0
US                 21.0
DL                 21.0
CO                 22.0
NW                 22.0
FL                 23.0
OO                 25.0
MQ                 26.0
9E                 27.0
EV                 27.0
AA                 28.0
XE                 28.0
OH                 30.0
UA                 30.0
B6                 32.0
YV                 33.0

Analitzant la mitjana dels minuts acumulats entre les hores de sortida estimades i reals per aerolínia, podriem determinar quina de les aerolínies és mes propensa a patir retrasos. Observem que la aerolínia WN que en l'analisi preliminar presentava major retràs acumulat, te un retràs mitjà de 20 minuts, i que per tant no sería la aerolinia que presenta mes retrasos en l'hora de sortida como podria interpretarse inicialment.

-------------------------------------------------------------------------------------------------------------------------------

In [15]:
# Troba quantes dades faltants hi ha per columna.

DelayedFlights.isnull().sum(axis=0)


Unnamed: 0.1             0
Unnamed: 0               0
Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                771
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime      901
CRSElapsedTime          19
AirTime                901
ArrDelay               901
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 771
TaxiOut                 33
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71300
WeatherDelay         71300
NASDelay             71300
SecurityDelay        71300
LateAircraftDelay    71300
dtype: int64

In [16]:
# Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).

# Nova columna "Velocitat Mitjana" en "miles / hour"

import pandas as pd
import numpy as np

DelayedFlights = pd.read_csv('DelayedFlights_Reduit.csv')

DelayedFlights['velocitat_mitjana'] = DelayedFlights['Distance'] / (DelayedFlights['AirTime'] / 60)

pd.options.display.max_columns = None

DelayedFlights

Unnamed: 0.1  Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  \
0             224054      711973  2008      2          18          1    734.0   
1            1775762     6562647  2008     12          23          2    901.0   
2            1095290     3643711  2008      7          14          1   1202.0   
3             185663      610669  2008      2           4          1   1620.0   
4             173677      566387  2008      1           3          4     40.0   
...              ...         ...   ...    ...         ...        ...      ...   
199995        160883      521433  2008      1          31          4   1257.0   
199996        559233     1749227  2008      3          17          1   1121.0   
199997       1087354     3620401  2008      7           7          1   2112.0   
199998        977672     3274068  2008      6          16          1   2235.0   
199999        359589     1130799  2008      2          24          7   1024.0   

        CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0              720   1040.0        1047            XE       2894  N14920   
1              835   1014.0         955            XE       2941  N12934   
2             1120   1212.0        1140            WN       1912  N349SW   
3             1525   2337.0        2255            WN        446  N237WN   
4             2215    857.0         629            CO        703  N73299   
...            ...      ...         ...           ...        ...     ...   
199995        1250   1542.0        1515            AA       1971  N459AA   
199996        1115   1414.0        1420            CO        438  N14653   
199997        2100   2216.0        2210            WN       3207  N613SW   
199998        2220   2336.0        2322            US       1788  N454UW   
199999        1015   1316.0        1303            B6        540  N562JB   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
0                   126.0           147.0    113.0      -7.0      14.0    MSY   
1                    73.0            80.0     55.0      19.0      26.0    IAH   
2                    70.0            80.0     57.0      32.0      42.0    SLC   
3                   257.0           270.0    236.0      42.0      55.0    LAS   
4                   317.0           314.0    290.0     148.0     145.0    SJC   
...                   ...             ...      ...       ...       ...    ...   
199995              225.0           205.0    198.0      27.0       7.0    DCA   
199996              173.0           185.0    144.0      -6.0       6.0    EWR   
199997               64.0            70.0     53.0       6.0      12.0    STL   
199998              121.0           122.0     99.0      14.0      15.0    CLT   
199999              172.0           168.0    141.0      13.0       9.0    PBI   

       Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  \
0       CLE       917     6.0      7.0          0                N         0   
1       BRO       308     4.0     14.0          0                N         0   
2       LAS       368     4.0      9.0          0                N         0   
3       ORF      2155     5.0     16.0          0                N         0   
4       EWR      2548    12.0     15.0          0                N         0   
...     ...       ...     ...      ...        ...              ...       ...   
199995  DFW      1192     8.0     19.0          0                N         0   
199996  MIA      1086     7.0     22.0          0                N         0   
199997  TUL       351     3.0      8.0          0                N         0   
199998  ORD       599     5.0     17.0          0                N         0   
199999  EWR      1024     8.0     23.0          0                N         0   

        CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                NaN           NaN       NaN            NaN   
1               19.0           0.0       0.0            0.0   

In [17]:
# Nova columna "Retard?"

    # Si l'hora d'arribada actual és major a l'hora d'arribada planificada, es produeix retard
        # Si ArrTime > CRSArrTime hi ha retard


DelayedFlights['Retard?'] = np.where(DelayedFlights['ArrTime'] > DelayedFlights['CRSArrTime'], 'Sí', 'No')

pd.options.display.max_columns = None

DelayedFlights



Unnamed: 0.1  Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  \
0             224054      711973  2008      2          18          1    734.0   
1            1775762     6562647  2008     12          23          2    901.0   
2            1095290     3643711  2008      7          14          1   1202.0   
3             185663      610669  2008      2           4          1   1620.0   
4             173677      566387  2008      1           3          4     40.0   
...              ...         ...   ...    ...         ...        ...      ...   
199995        160883      521433  2008      1          31          4   1257.0   
199996        559233     1749227  2008      3          17          1   1121.0   
199997       1087354     3620401  2008      7           7          1   2112.0   
199998        977672     3274068  2008      6          16          1   2235.0   
199999        359589     1130799  2008      2          24          7   1024.0   

        CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  FlightNum TailNum  \
0              720   1040.0        1047            XE       2894  N14920   
1              835   1014.0         955            XE       2941  N12934   
2             1120   1212.0        1140            WN       1912  N349SW   
3             1525   2337.0        2255            WN        446  N237WN   
4             2215    857.0         629            CO        703  N73299   
...            ...      ...         ...           ...        ...     ...   
199995        1250   1542.0        1515            AA       1971  N459AA   
199996        1115   1414.0        1420            CO        438  N14653   
199997        2100   2216.0        2210            WN       3207  N613SW   
199998        2220   2336.0        2322            US       1788  N454UW   
199999        1015   1316.0        1303            B6        540  N562JB   

        ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin  \
0                   126.0           147.0    113.0      -7.0      14.0    MSY   
1                    73.0            80.0     55.0      19.0      26.0    IAH   
2                    70.0            80.0     57.0      32.0      42.0    SLC   
3                   257.0           270.0    236.0      42.0      55.0    LAS   
4                   317.0           314.0    290.0     148.0     145.0    SJC   
...                   ...             ...      ...       ...       ...    ...   
199995              225.0           205.0    198.0      27.0       7.0    DCA   
199996              173.0           185.0    144.0      -6.0       6.0    EWR   
199997               64.0            70.0     53.0       6.0      12.0    STL   
199998              121.0           122.0     99.0      14.0      15.0    CLT   
199999              172.0           168.0    141.0      13.0       9.0    PBI   

       Dest  Distance  TaxiIn  TaxiOut  Cancelled CancellationCode  Diverted  \
0       CLE       917     6.0      7.0          0                N         0   
1       BRO       308     4.0     14.0          0                N         0   
2       LAS       368     4.0      9.0          0                N         0   
3       ORF      2155     5.0     16.0          0                N         0   
4       EWR      2548    12.0     15.0          0                N         0   
...     ...       ...     ...      ...        ...              ...       ...   
199995  DFW      1192     8.0     19.0          0                N         0   
199996  MIA      1086     7.0     22.0          0                N         0   
199997  TUL       351     3.0      8.0          0                N         0   
199998  ORD       599     5.0     17.0          0                N         0   
199999  EWR      1024     8.0     23.0          0                N         0   

        CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0                NaN           NaN       NaN            NaN   
1               19.0           0.0       0.0            0.0   

In [19]:
# Salvem el nou Dataframe amb les noves columnes com a dataset "DelayedFlights_Reduit_Columnes.csv"

DelayedFlights.to_csv('DelayedFlights_Reduit_Columnes.csv')

In [20]:
# Fes una taula de les aerolínies amb més endarreriments acumulats.

    # Prenem les variables:
        #"UniqueCarrier" (aerolínia)
        # "ArrDelay" (Diferencia en minuts entre l'hora estimada d'arribada i la real (si el vol arriba abans de l'hora estimada, número negatiu en minuts)
        
    # Ordenem taula en ordre descendent: primer les aerolínes amb major retard acumulat en l'arribada

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

Endarreriments_aerolínies = DelayedFlights[['UniqueCarrier','ArrDelay']]

Endarreriments_aerolínies.groupby('UniqueCarrier').sum().sort_values('ArrDelay', ascending = False)



ArrDelay
UniqueCarrier           
WN             1161056.0
AA              921288.0
UA              691318.0
MQ              664277.0
OO              616109.0
XE              525118.0
DL              465143.0
CO              426629.0
EV              387220.0
US              384366.0
YV              383519.0
NW              344087.0
FL              324040.0
B6              312811.0
OH              265159.0
9E              259339.0
AS              140842.0
F9               81701.0
HA               22560.0
AQ                2256.0

In [21]:
# Quins són els vols més llargs?

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

Vols_mes_llargs = DelayedFlights[['UniqueCarrier','FlightNum', 'Origin', 'Dest', 'Distance']]

Vols_mes_llargs.sort_values('Distance', ascending = False)

# ¿Posible Alternativa?:  Vols_mes_llargs.groupby(['UniqueCarrier','FlightNum', 'Origin', 'Dest']).head().sort_values('Distance', ascending = False)

UniqueCarrier  FlightNum Origin Dest  Distance
69860             CO         15    EWR  HNL      4962
65169             CO         15    EWR  HNL      4962
84241             CO         15    EWR  HNL      4962
96690             CO         15    EWR  HNL      4962
116259            CO         14    HNL  EWR      4962
...              ...        ...    ...  ...       ...
62008             AS         65    WRG  PSG        31
23487             AS         65    WRG  PSG        31
54021             AS         65    WRG  PSG        31
23129             AS         64    PSG  WRG        31
126761            OH       5572    JFK  LGA        11

[200000 rows x 5 columns]

In [22]:
#  I els més endarrerits?

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

Vols_mes_endarrerits = DelayedFlights[['UniqueCarrier', 'FlightNum', 'Origin', 'Dest', 'ArrDelay']]

Vols_mes_endarrerits.sort_values('ArrDelay', ascending = False)


UniqueCarrier  FlightNum Origin Dest  ArrDelay
128644            AA       2228    HDN  DFW    1279.0
103492            NW        804    KOA  SEA    1202.0
188146            MQ       4801    CVG  LGA    1195.0
57765             NW        776    LAS  MSP    1190.0
187908            AA       1280    DEN  ORD    1143.0
...              ...        ...    ...  ...       ...
198842            DL        792    ATL  ORD       NaN
199596            OO       5857    ORD  PIA       NaN
199600            B6        477    BOS  OAK       NaN
199641            AA        384    DFW  JFK       NaN
199739            CO        596    LAS  IAH       NaN

[200000 rows x 5 columns]

In [33]:
# Busca les rutes més llargues

import pandas as pd
import numpy as np

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

# obtenció de les 5 rutes més llargues

 # DelayedFlights.nlargest(5,[('Distance')]).DelayedFlights['Distance'].unique()

DelayedFlights.drop_duplicates('Distance').nlargest(5,[('Distance')])

Unnamed: 0.2  Unnamed: 0.1  Unnamed: 0  Year  Month  DayofMonth  \
4104           4104       1519768     5360481  2008      9          13   
6939           6939        727247     2380692  2008      4          10   
42               42       1809922     6660910  2008     12           1   
51082         51082        956584     3214439  2008      6           1   
32419         32419         66627      218294  2008      1           4   

       DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
4104           6   2004.0        1955   1128.0        1130            CO   
6939           4   1025.0        1010   1433.0        1355            DL   
42             1   1023.0        1008   1521.0        1520            UA   
51082          7   1651.0        1645    546.0         541            UA   
32419          5   1028.0         954   1545.0        1501            UA   

       FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  \
4104          14  N66056              564.0           575.0    541.0   
6939        1273  N828MH              608.0           585.0    571.0   
42             1  N214UA              538.0           552.0    510.0   
51082          3  N210UA              475.0           476.0    461.0   
32419          3  N215UA              557.0           547.0    516.0   

       ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  Cancelled  \
4104       -2.0       9.0    HNL  EWR      4962     8.0     15.0          0   
6939       38.0      15.0    ATL  HNL      4502     4.0     33.0          0   
42          1.0      15.0    ORD  HNL      4243     2.0     26.0          0   
51082       5.0       6.0    KOA  ORD      4213     5.0      9.0          0   
32419      44.0      34.0    ORD  OGG      4184    21.0     20.0          0   

      CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
4104                 N         0           NaN           NaN       NaN   
6939                 N         0          11.0           0.0      23.0   
42                   N         0           NaN           NaN       NaN   
51082                N         0           NaN           NaN       NaN   
32419                N         0          34.0           0.0      10.0   

       SecurityDelay  LateAircraftDelay  velocitat_mitjana Retard?  
4104             NaN                NaN         550.314233      No  
6939             0.0                4.0         473.064799      Sí  
42               NaN                NaN         499.176471      Sí  
51082            NaN                NaN         548.329718      Sí  
32419            0.0                0.0         486.511628      Sí

Les 5 rutes mes llargues son:
- HNL - EWR: 4962 miles
- ATL - HNL: 4502 miles
- ORD - HNL: 4243 miles
- KOA - ORD: 4213 miles
- ORD - OGG: 4184 miles

In [48]:
# les rutes que acumulen més retards

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

# obtenció de les 5 rutes amb més retards

DelayedFlights[['Origin','Dest','ArrDelay']].nlargest(5,['ArrDelay'])




Origin Dest  ArrDelay
128644    HDN  DFW    1279.0
103492    KOA  SEA    1202.0
188146    CVG  LGA    1195.0
57765     LAS  MSP    1190.0
187908    DEN  ORD    1143.0

------------------------------------------------------------------------------------------------------------

Exercici 3

In [49]:
# Exporta el dataset net i amb les noves columnes a Excel

import pandas as pd

DelayedFlights = pd.read_csv('DelayedFlights_Reduit_Columnes.csv')

DelayedFlights.to_excel('Dataset_Net_Columnes.xlsx', index=False)